### PR TITLE
PoSAccounting simulation test

### LINKED ISSUE #568: Add simulation test for PoSAccountingOperations
Test must perform random operations and check invariants. For example that owner balance + all delegations balance = pool balance

### PR DESCRIPTION
**Pull request by azarovh**
Add tests that perform random operations and check storage consistency on every iteration.

Closes #568

### CODE REVIEW

File: pos_accounting/src/storage/in_memory.rs [hunk 1]
```diff
@@ -59,6 +59,69 @@ impl InMemoryPoSAccounting {
...
+        // pool_balance and pool_data must contain the same keys
+        assert_eq!(self.pool_balances.len(), self.pool_data.len());
+        let pool_balance_and_data_consistant =
+            self.pool_balances.keys().all(|key| self.pool_data.contains_key(key));
+        assert!(pool_balance_and_data_consistant);
+
+        // delegation_balance and delegation_data must contain the same keys
+        //
+        // Note: delegation_balances and delegation_data can have different length
+        // because zero balances are removed.
```
Comment (azarovh) [inline pos_accounting/src/storage/in_memory.rs:89]: Initially, we decided not to store zero balances in deltas and storage because it seemed redundant. But maybe we should reconsider this for consistency reasons as it would be easier to check the state. Thoughts?

File: pos_accounting/src/pool/storage/operator_impls.rs [hunk 2]
```diff
@@ -301,7 +310,10 @@ impl<S: PoSAccountingStorageWrite<T>, T: StorageTag> PoSAccountingDB<S, T> {
 
         self.add_to_delegation_balance(undo_data.delegation_id, undo_data.amount)?;
 
-        self.add_balance_to_pool(pool_id, undo_data.amount)?;
+        // it's possible that the pool was decommissioned
+        if self.pool_exists(pool_id)? {
+            self.add_balance_to_pool(pool_id, undo_data.amount)?;
+        }
```
Comment (iljakuklic) [inline pos_accounting/src/pool/storage/operator_impls.rs:316]: Is it legal to have a non-existing pool here? I.e. should there be an `else` with an error?
This is replicated in multiple places, could the code somehow be shared?

File: pos_accounting/src/storage/in_memory.rs [hunk 3]
```diff
@@ -59,6 +59,69 @@ impl InMemoryPoSAccounting {
...
+        // delegation balances and delegation shares must contain the same amounts
+        assert_eq!(
+            self.delegation_balances.len(),
+            self.pool_delegation_shares.len()
+        );
+        let delegation_shares_and_balances_consistent = self
+            .pool_delegation_shares
+            .iter()
+            .all(|((_, key), balance)| self.delegation_balances.get(key) == Some(balance));
+        assert!(delegation_shares_and_balances_consistent);
```
Comment (iljakuklic) [inline pos_accounting/src/storage/in_memory.rs:123]: Wondering if there's a value in using `.for_each(... assert_eq!` instead of `.all(... ==`. You get information about what failed rather than that something failed.

**Review by iljakuklic — APPROVED** [review]
Nice. Some minor comments below.

File: pos_accounting/src/pool/storage/operator_impls.rs (hunk 2 above)
Comment (azarovh) [reply pos_accounting/src/pool/storage/operator_impls.rs:316]: It is legal. The reason is that decommissioning pool removes only pool balance and data. Pool's owner cannot affect delegations, so they are kept in the db until delegation owner decides to spend it.

File: pos_accounting/src/storage/in_memory.rs (hunk 3 above)
Comment (azarovh) [reply pos_accounting/src/storage/in_memory.rs:123]: makes sense, will do

File: pos_accounting/src/pool/storage/operator_impls.rs [hunk 4]
```diff
@@ -301,7 +310,10 @@ impl<S: PoSAccountingStorageWrite<T>, T: StorageTag> PoSAccountingDB<S, T> {
 
         self.add_to_delegation_balance(undo_data.delegation_id, undo_data.amount)?;
 
-        self.add_balance_to_pool(pool_id, undo_data.amount)?;
+        // it's possible that the pool was decommissioned
```
Comment (TheQuantumPhysicist) [inline pos_accounting/src/pool/storage/operator_impls.rs:313]: Decommissioning the pool should not prevent people from withdrawing their staked amounts.

File: pos_accounting/src/storage/in_memory.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply pos_accounting/src/storage/in_memory.rs:93]: What's wrong with "if it's not availabe, it's zero"?

Comment (azarovh) [reply pos_accounting/src/storage/in_memory.rs:93]: It's impossible to know if it's "not available because it's zero" or "not available because of some error"

Comment (TheQuantumPhysicist) [reply pos_accounting/src/storage/in_memory.rs:93]: The problem there is that we'll have these zeroes forever. Not a good deal.

What you're talking about is not an error, but an inconsistency. 

So we have to store tons of useless data because we can't do proper arithmetic. I think we can do proper arithmetic checks to ensure that non-existent is zero.

File: pos_accounting/src/pool/storage/operator_impls.rs (hunk 4 above)
Comment (TheQuantumPhysicist) [reply pos_accounting/src/pool/storage/operator_impls.rs:313]: I'm trying to think if there's a corner case that isn't covered where this can be exploited.

File: pos_accounting/src/pool/tests/simulation_tests.rs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+fn perform_random_operation(
+    rng: &mut (impl Rng + CryptoRng),
+    op: &mut (impl PoSAccountingOperations + PoSAccountingView),
+    undos: &mut Vec<PoSAccountingUndo>,
+    random_pool: Option<PoolId>,
+    random_delegation: Option<(DelegationId, DelegationData)>,
+    random_delegation_balance: Option<(DelegationId, Amount)>,
+) {
+    match rng.gen_range(0..11) {
+        0 | 1 => {
```
Comment (TheQuantumPhysicist) [inline pos_accounting/src/pool/tests/simulation_tests.rs:151]: Please label every arm with a comment on what's going on inside of it.

File: pos_accounting/src/pool/tests/simulation_tests.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            let (_, undo) = op.create_pool(&input0_outpoint, pool_data).unwrap();
+            undos.push(undo);
+        }
+        2 => {
+            if let Some(pool_id) = random_pool {
+                let undo = op.decommission_pool(pool_id).unwrap();
+                undos.push(undo);
+            }
+        }
+        3 | 4 => {
```
Comment (TheQuantumPhysicist) [inline pos_accounting/src/pool/tests/simulation_tests.rs:165]: You can write `3..=4`... which is more readable.

File: pos_accounting/src/pool/tests/simulation_tests.rs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+fn perform_random_operation(
+    rng: &mut (impl Rng + CryptoRng),
+    op: &mut (impl PoSAccountingOperations + PoSAccountingView),
+    undos: &mut Vec<PoSAccountingUndo>,
+    random_pool: Option<PoolId>,
+    random_delegation: Option<(DelegationId, DelegationData)>,
+    random_delegation_balance: Option<(DelegationId, Amount)>,
+) {
+    match rng.gen_range(0..11) {
```
Comment (TheQuantumPhysicist) [inline pos_accounting/src/pool/tests/simulation_tests.rs:150]: One of my concerns that I've been thinking about is, what if we, in the future, add new operations? How are we gonna remember to add them to the tests?

Here's a solution: There's a crate named `variant_count`. With it, you can do:
```
assert_eq!(PoSAccountingUndo::VARIANT_COUNT, 6);
```

This way, if new ops are added, this assert will fire. With this assert, you can write a comment/message explaining what's going on and why.

Comment (TheQuantumPhysicist) [reply pos_accounting/src/pool/tests/simulation_tests.rs:150]: This match can be done in a much nicer way... using proper arms and probabilities... but I'm not gonna fixate on that now.

**Review by TheQuantumPhysicist — APPROVED** [review]
I'm approving this, but I'm concerned about one thing: Popping things randomly from database basically means order doesn't matter. But is this true?

File: pos_accounting/src/pool/tests/simulation_tests.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+fn get_random_pool_id(rng: &mut impl Rng, storage: &InMemoryPoSAccounting) -> Option<PoolId> {
+    let all_pool_data = storage.all_pool_data();
+    if !all_pool_data.is_empty() {
+        let (pool_id, _) = all_pool_data.iter().nth(rng.gen_range(0..all_pool_data.len())).unwrap();
+        Some(*pool_id)
+    } else {
+        None
+    }
```
Comment (iljakuklic) [inline pos_accounting/src/pool/tests/simulation_tests.rs:45]: Not sure it's much of an improvement but it seems you can use [bool::then](https://github.com/Cameri/nostream) here and a few other functions in this file:
```rust
(!all_pool_data.is_empty()).then(|| all_pool_data.iter().blah...())
```

**Review by iljakuklic — APPROVED** [review]
+1

**Comment by TheQuantumPhysicist** [comment]
>> Not sure that I understand your concern, isn't it what happens in production? Users doing random actions. Btw everything is random but `undo` is always for the latest action.  

Random, but they're in order because of utxos.
